refactor: Изменил фильтры и добавил поиск

Denis-Shtanskiy opened the pull request:
- Исправил баги обнаруженные тестировщиками
- Провел рефакторинг фильтрации (сделал как у Иры)
- Добавил полнотекстовый поиск

Denis-Shtanskiy committed:
refactor: Изменил фильтры и добавли поиск
- Исправил баги обнаруженные тестировщиками
- Добавил ограничения по фильтрации (сделал как у Иры)
- Добавил поиск

Denis-Shtanskiy committed:
Merge branch 'develop' into feature/search

Denis-Shtanskiy committed:
fix: Пулл изменений из девелоп
- Забыл о возможностях создавать проекты и описания на разных языках

Badmajor commented on src/backend/api/v1/projects/filters.py:
```diff
@@ -62,6 +73,17 @@ def get_project_role(self, queryset, name, value):
+            vector_ru = SearchVector("name", "description", config="russian")
+            vector_en = SearchVector("name", "description", config="english")
+            return queryset.annotate(search=vector_ru | vector_en).filter(
+                search=search_query_ru | search_query_en
```
Нужно ли тут запрашивать два кверисета с реультатами, может лучше сначала опрелеить язык, например с помощью библиотеки langdetect, и вместо двух запросов делать один?

Denis-Shtanskiy replied:
Посмотрю, не знал про библиотеку, спасибо) Думаю да, так вернее будет, через ветвление определить язык и тд

Badmajor reviewed (commented):
Вопрос к def project_search

ArtemKAF commented on src/backend/apps/projects/migrations/0007_project_search_vector_and_more.py:
В develop было 8 файлов миграций. Теперь 9.

ArtemKAF commented on src/backend/apps/projects/models.py:
https://github.com/Pet-projects-CodePET/Backend/blob/d6756fb7b38b0849dd3b086d1e24cc467dd9fef1/src/backend/apps/projects/models.py#L163-L165
Действительно достаточно без указания полей, на основе которых будет строиться вектор?

ArtemKAF commented on src/backend/api/v1/projects/filters.py:
https://github.com/Pet-projects-CodePET/Backend/blob/d6756fb7b38b0849dd3b086d1e24cc467dd9fef1/src/backend/api/v1/projects/filters.py#L76-L85
```shell
File "/home/artemkaf/Dev/CodePET-Backend/src/backend/api/v1/projects/filters.py", line 82, in project_search
    return queryset.annotate(search=vector_ru | vector_en).filter(
                                    ~~~~~~~~~~^~~~~~~~~~~
  File "/home/artemkaf/.cache/pypoetry/virtualenvs/codepet-backend-UEAheHnO-py3.12/lib/python3.12/site-packages/django/db/models/expressions.py", line 124, in __or__
    raise NotImplementedError(
NotImplementedError: Use .bitand(), .bitor(), and .bitxor() for bitwise logical operations.
```
При попытке воспользоваться поиском следующим запросом:
```
GET http://localhost:9000/api/v1/projects/?search=test
```

ArtemKAF reviewed (approved):
Будет интересно посмотреть как это будет работать. Локально воспроизвести рабочий вариант не удалось.

Denis-Shtanskiy committed:
Merge branch 'develop' into feature/search

Denis-Shtanskiy committed:
fix: Исправил замечания
- Сделал мердж по конфликтующим миграциям
- Добавил библиотеку langdetect в зависимости тоже
- Чуть-чуть доработал локальный докер-композ файл, чтобы переменные брались из корневой директории

Denis-Shtanskiy committed:
fix: Исправил замечания
- Сделал мердж по конфликтующим миграциям
- Добавил библиотеку langdetect в зависимости тоже
- Чуть-чуть доработал локальный докер-композ файл, чтобы переменные брались из корневой директории

ArtemKAF commented on src/backend/apps/projects/models.py:
В данной реализации, что поле SearchVectorField, что индекс GinIndex добавлены бесполезно. Они никак не используются в процессе поиска.

Denis-Shtanskiy replied:
Убрал поля

ArtemKAF reviewed (changes requested):
Нужно, либо доработать так, чтобы поле в модели с вектором поиска и индексом стали полезными, либо убрать их. На данный момент только лишние миграции от их добавления.

RavenIV reviewed (commented):
Согласна с Артемом, в этом варианте добавление SearchVectorField и GinIndex выглядит бессмысленно. Так понимаю, чтобы SearchVectorField работал, его нужно в каком-то месте заполнить триггерами, как в доке [написано](https://docs.djangoproject.com/en/5.0/ref/contrib/postgres/search/#:~:text=text%20search.-,SearchVectorField,-%C2%B6). 
Вообще тема поиска очень интересная, будет здорово послушать, что успели накопать)

Denis-Shtanskiy committed:
fix: Убрал поля, смерджил девелоп

Denis-Shtanskiy committed:
fix:update poetry.lock
